[MacCoss] Panorama Public Search Improvements - Submit upon Enter key press, clear input and reset, add search criteria, etc

labkey-bpatel opened the pull request:
#### Rationale
Additional changes/fixes for requested by the client, and other updates.

Spec/Test plan:  https://docs.google.com/document/d/1zczm_0IORm7ci-OfCdOiGuMo9x-eRv2h30RB6BDBZak/edit#

#### Changes
* Submit form within a tab by pressing Enter key
* Remove custom search indicator, and rely on built in DataRegion search indicator ('loading..').
* Set container filter on protein / peptide search results to 'Current and Subfolders'
* Experiment, Protein and Peptide search - Show search results outside of search panel in an existing Targeted MS Experiment List webpart. Handle filtering, refresh, url update
* Add search criteria for Protein and Peptide results
* Implement Clear All to clear the inputs and reset webpart to 'Targeted MS Experiment List' webpart
* Handle search when input fields are blank

labkey-bpatel committed:
Make Author and Title 'tags' fields. Add 'Clear All' to clear input fields. Add search criteria.

labkey-bpatel committed:
Remove 'Targeted MS Experiment List' webpart upon Experiment search

labkey-bpatel committed:
Submitting form within a tab by pressing Enter Key

labkey-bpatel committed:
minor reformat

labkey-bpatel committed:
Set contailer filter on protein / peptide search results to 'Current and Subfolders', so that clicking the "View" button on the result grid displays expected results

labkey-bpatel committed:
Handle Organism value getting truncated from the URL, update search criteria and title string format

labkey-bpatel committed:
Experiment search - Show search results outside of search panel in an existing Targeted MS Experiment List webpart. Handle filtering, and refresh.

labkey-bpatel committed:
Protein and Peptide search results - Show results outside of search panel in an existing Targeted MS Experiment List webpart. Handle filtering, and refresh.

labkey-bpatel committed:
Reset url upon Clearing all inputs

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/search/panoramaWebSearch.jsp:
```diff
@@ -82,6 +82,10 @@
                     </td>
+                    <td style="width: 5px"></td>
+                    <td>
+                        <button id="clear-all-button-id-experiment" class="clear-all-button" onclick=clearInputFieldsAndResetURL("experiment")>Clear All</button>
```
The onclick handlers' HTML is malformed. I know it works, but it should be more like:

```suggestion
                        <button id="clear-all-button-id-experiment" class="clear-all-button" onclick="clearInputFieldsAndResetURL('experiment')">Clear All</button>
```

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/search/panoramaWebSearch.jsp:
```diff
@@ -187,6 +197,25 @@
     });
 
+    //submit form via Enter key
+    $('.tab-1').keypress((e) => {
```
You could have a single function that's used for all three of these instead of three copies.

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/search/panoramaWebSearch.jsp:
```diff
@@ -222,11 +257,31 @@
+        switch (searchTabName) {
+            case "experiment":
+                clearInputFromExperimentTab();
+                resetUrl();
```
Can move `resetUrl()` outside of the `switch` since it's identical across all branches.

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/search/panoramaWebSearch.jsp:
```diff
@@ -244,164 +299,205 @@
+                        if (expWebpart.length === 1) {
+                            let wp = new LABKEY.QueryWebPart({
+                                renderTo: 'webpart_'+ expWebpart[0].webPartId,
+                                title: 'Targeted MS Experiment List',
```
Let's title this 'Panorama Public Experiments' or something like that.

labkey-jeckels commented on panoramapublic/webapp/PanoramaPublic/css/PanoramaWebSearch.css:
```diff
@@ -74,4 +74,16 @@ input::placeholder {
+    border: none;
+    padding: 0!important;
+    /*optional*/
+    font-family: "Roboto",Arial,Helvetica,sans-serif;
```
Why does Roboto get quotes but other fonts don't?

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/search/panoramaWebSearch.jsp:
```diff
@@ -244,164 +299,205 @@
                 if (author) {
                     expAnnotationFilters.push(createFilter(authorsItemId, author));
-                    updateUrlFilters(null, authorsItemId, author);
+                    expSearchParams += "Targeted MS Experiment List." + "author~containsoneof" + "=" + author + "&";
```
`author` should be URI-encoded

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/search/panoramaWebSearch.jsp:
```diff
@@ -244,164 +299,205 @@
             if (context[exactPeptideMatchesItemId]) {
                 peptideParameters[exactMatch] =  context[exactPeptideMatchesItemId];
-                document.getElementById(exactPeptideMatchesItemId).value = context[exactPeptideMatchesItemId];
+                context[exactPeptideMatchesItemId] === "true" ? (document.getElementById(exactPeptideMatchesItemId).checked = true): (document.getElementById(exactPeptideMatchesItemId).checked = false);
```
```suggestion
                document.getElementById(exactPeptideMatchesItemId).checked = context[exactPeptideMatchesItemId] === "true";
```

labkey-bpatel committed:
Code review changes, and bug fixes:
- fix malformed html
- refactor to use single function for submit with enter key
- move resetUrl() to outside of switch
- Experiment search result title change to 'Panorama Public Experiments'
- uri encode
- updates to handle search when input fields are blank
- fix url so that it only loads page when Search is clicked or Enter key is pressed

labkey-bpatel committed:
- use 'authors' instead of 'author'

labkey-bpatel committed:
- get index of '#' instead of '?' for resetUrl

labkey-bpatel committed:
Experiment search test update

labkey-bpatel committed:
Update protein search auto test

labkey-bpatel committed:
Update peptide search auto test, fix regressions

labkey-bpatel committed:
test fixes

labkey-bpatel committed:
minor cleanup

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/search/panoramaWebSearch.jsp:
```diff
@@ -244,164 +286,205 @@
                 let organism = document.getElementById(organismItemId).value;
                 let instrument = document.getElementById(instrumentItemId).value;
 
+                let expSearchParams = "";
                 if (author) {
                     expAnnotationFilters.push(createFilter(authorsItemId, author));
-                    updateUrlFilters(null, authorsItemId, author);
+                    expSearchParams += "Targeted MS Experiment List." + "authors~containsoneof" + "=" + author + "&";
                 }
                 if (title) {
                     expAnnotationFilters.push(createFilter(titleItemId, title));
-                    updateUrlFilters(null, titleItemId, title);
+                    expSearchParams += "Targeted MS Experiment List." + "title~containsoneof" + "=" + title + "&";
                 }
                 if (organism) {
                     expAnnotationFilters.push(createFilter(organismItemId, organism));
-                    updateUrlFilters(null, organismItemId, organism);
+                    expSearchParams += "Targeted MS Experiment List." + "organism~containsoneof" + "=" + organism + "&";
                 }
                 if (instrument) {
                     expAnnotationFilters.push(createFilter(instrumentItemId, instrument));
-                    updateUrlFilters(null, instrumentItemId, instrument);
+                    expSearchParams += "Targeted MS Experiment List." + "instrument~containsoneof" + "=" + instrument;
+                }
+                if (expSearchParams !== "") {
+                    location.replace(window.location.href + "?" + expSearchParams);
```
These values still need to be URI-encoded as they're being appended.

labkey-bpatel replied:
Yes! I thought i had done that, must have reverted it somehow with my refactors
Now done, again

labkey-jeckels reviewed (approved):
I'll approve but please be sure to URI encode the values as you append them.

labkey-bpatel committed:
auto test fix - scroll into view the second row

labkey-bpatel committed:
Encode uri

labkey-bpatel committed:
auto test fix - scroll results into view

labkey-bpatel committed:
Update to handle free text along with the value from dropdown list for Organism and Instrument fields

labkey-sweta reviewed (changes requested):
There is method search which returns the updated dataRegionTable from latest change, any particular reason for new clickSearch method?

labkey-sweta committed:
Stale element exception fix.

labkey-sweta committed:
test code clean up

labkey-sweta committed:
Adding test coverage for empty protein and peptide search.

labkey-sweta committed:
Clean up

labkey-sweta committed:
Rollback some changes

labkey-bpatel committed:
Test fix

labkey-bpatel committed:
Test fix - find data region again after each search

labkey-bpatel committed:
Test fix attempt, some code reformat, fix refresh on Experiment Search

labkey-bpatel committed:
Wait for element

labkey-bpatel committed:
find the data region by its name

labkey-bpatel committed:
auto test fix - try wrapping click event with ElementCache

labkey-bpatel committed:
auto test fix - wait for rows

labkey-bpatel commented:
Panorama web search related tests are now passing, merging...
